tagging/apply_tags.py

```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env python3
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+import argparse
+import logging
+from tagger_interface import TaggerInterface
+from dataclasses import dataclass, field
+from typing import Optional, List
+from taggers import SHATagger, PythonVersionTagger
+
+
+# NB_VERSION_TAG="notebook-$(docker run --rm -a STDOUT ${IMAGE_NAME} jupyter-notebook --version | tr -d '\r')"
+# docker tag $IMAGE_NAME "$DOCKER_REPO:${NB_VERSION_TAG%% }"
+# LAB_VERSION_TAG="lab-$(docker run --rm -a STDOUT ${IMAGE_NAME} jupyter-lab --version | tr -d '\r')"
+# docker tag $IMAGE_NAME "$DOCKER_REPO:${LAB_VERSION_TAG%%\r}"
+# HUB_VERSION_TAG="hub-$(docker run --rm -a STDOUT ${IMAGE_NAME} jupyterhub --version | tr -d '\r')"
+# docker tag $IMAGE_NAME "$DOCKER_REPO:${HUB_VERSION_TAG%%\r}"
+
+
+@dataclass
+class ImageDescription:
+    parent_image: Optional[str]
+    taggers: List[TaggerInterface] = field(default_factory=list)
+
+
+ALL_IMAGES = {
+    "minimal-notebook": ImageDescription(
+        parent_image=None,
+        taggers=[SHATagger]
+    ),
+    "base-notebook": ImageDescription(
+        parent_image="minimal-notebook",
+        taggers=[PythonVersionTagger]
+    )
+}
+
+
+def get_all_taggers(short_image_name):
+    taggers = []
+    while short_image_name is not None:
+        image_description = ALL_IMAGES[short_image_name]
+        taggers = image_description.taggers + taggers
+        short_image_name = image_description.parent_image
+    return taggers
```

Comment:
So, we're inheriting all the taggers of the base images, but not the values of tags.

We are going to reexecute the taggers on image, because this will be helpful for manifest files (python packages are changing between different, so if we use taggers for manifest file creation, we have to reexecute the taggers).

Reply:
Nice!